pkgs/build-support/rust/fetch-cargo-vendor-util.py

```diff
@@ -0,0 +1,284 @@
+    if pkg["source"] != "registry+https://github.com/rust-lang/crates.io-index":
+        raise Exception("Only the default crates.io registry is supported.")
+
+    return f"https://crates.io/api/v1/crates/{pkg["name"]}/{pkg["version"]}/download"
+
+
+def download_tarball(pkg: dict[str, Any], out_dir: Path) -> None:
+
+    url = get_download_url_for_tarball(pkg)
+    filename = f"{pkg["name"]}-{pkg["version"]}.tar.gz"
+
+    # TODO: allow legacy checksum specification, see importCargoLock for example
+    #       also, don't forget about the other usage of the checksum
+    expected_checksum = pkg["checksum"]
+
+    tarball_out_dir = out_dir / "tarballs" / filename
+    eprint(f"Fetching {url} -> tarballs/{filename}")
+
+    calculated_checksum = download_file_with_checksum(url, tarball_out_dir)
+
+    if calculated_checksum != expected_checksum:
+        raise Exception(f"Hash mismatch! File fetched from {url} had checksum {calculated_checksum}, expected {expected_checksum}.")
+
+
+def download_git_tree(url: str, git_sha_rev: str, out_dir: Path) -> None:
+
+    tree_out_dir = out_dir / "git" / git_sha_rev
+    eprint(f"Fetching {url}#{git_sha_rev} -> git/{git_sha_rev}")
+
+    cmd = ["nix-prefetch-git", "--builder", "--quiet", "--url", url, "--rev", git_sha_rev, "--out", str(tree_out_dir)]
+    subprocess.check_output(cmd)
+
+
+GIT_SOURCE_REGEX = re.compile("git\\+(?P<url>[^?]+)(\\?(?P<type>rev|tag|branch)=(?P<value>.*))?#(?P<git_sha_rev>.*)")
+
+
+class GitSourceInfo(TypedDict):
+    url: str
+    type: str | None
+    value: str | None
+    git_sha_rev: str
+
+
+def parse_git_source(source: str) -> GitSourceInfo:
+    match = GIT_SOURCE_REGEX.match(source)
+    if match is None:
+        raise Exception(f"Unable to process git source: {source}.")
+    return cast(GitSourceInfo, match.groupdict(default=None))
+
+
+def create_vendor_staging(lockfile_path: Path, out_dir: Path) -> None:
+    cargo_toml = load_toml(lockfile_path)
+
+    git_packages: list[dict[str, Any]] = []
+    registry_packages: list[dict[str, Any]] = []
+
+    for pkg in cargo_toml["package"]:
+        # ignore local dependenices
+        if "source" not in pkg.keys():
+            eprint(f"Skipping local dependency: {pkg["name"]}")
+            continue
+        source = pkg["source"]
+
+        if source.startswith("git+"):
+            git_packages.append(pkg)
+        elif source.startswith("registry+"):
+            registry_packages.append(pkg)
+        else:
+            raise Exception(f"Can't process source: {source}.")
+
+    git_sha_rev_to_url: dict[str, str] = {}
+    for pkg in git_packages:
+        source_info = parse_git_source(pkg["source"])
+        git_sha_rev_to_url[source_info["git_sha_rev"]] = source_info["url"]
+
+    out_dir.mkdir(exist_ok=True)
+    shutil.copy(lockfile_path, out_dir / "Cargo.lock")
+
+    # create a pool with at most 10 concurrent jobs
+    with mp.Pool(min(10, mp.cpu_count())) as pool:
```

Comment:
Since this is IO based, I don't think we need to bind this to the number of cores. 
10 is a bit more than what browsers do (firefox does 10). Otherwise when I run this on an EPYC, I get 96 connections, which github might not like.

```suggestion
    with mp.Pool(10) as pool:
```

Reply:
I used min(10,x) to clamp x to not be greater than 10.
I could change it to 7. Maybe I should look at cargo's implementation.

If we just use 10 or 7 without considering the core count will it know automatically not to try more things it can handle? I'll probably have to read the docs of multiprocessing to know for sure.

Reply:
I think `min` was misread as `max`.

Reply:
Looked up [the docs](https://docs.python.org/3/library/multiprocessing.html#module-multiprocessing.pool) and the default value for the number of processes seems to be [os.cpu_count()](https://docs.python.org/3/library/os.html#os.cpu_count) for `python<=3.12` and [os.process_cpu_count()](https://docs.python.org/3/library/os.html#os.process_cpu_count) for `python>=3.13`.

Based on this info, I would say that we should use the current clamping method.

(I'm assuming `multiprocessing.cpu_count()` is the same as `os.cpu_count()`)

Reply:
Ah, right. Mixed up the logic.